### ISSUE TITLE
Shebang in .ts (TypeScript) file causes it to render with .js (JavaScript) icon

### ISSUE THREAD

**Issue by sompylasar** [file-icons/atom#606]
<img width="493" alt="typescript with shebang renders javascript icon" src="https://cloud.githubusercontent.com/assets/498274/26534458/35f2e00a-43d9-11e7-8e38-dbf12406f1be.png">

<img width="488" alt="typescript without shebang renders typescript icon" src="https://cloud.githubusercontent.com/assets/498274/26534461/3a38e542-43d9-11e7-98fa-621daf62300e.png">

**Comment by Alhadis** [comment]
Well... yeah, that's the expected behaviour.

Are you saying Node can directly execute TypeScript files without precompilation?

**Comment by sompylasar** [comment]
No, it cannot, but the shebang string gets propagated from the TS source to the JS compiled with `tsc`. If I omit it in the source, it won't appear in the output JS.

**Comment by Alhadis** [comment]
Ahhh, I see. Do you know if TypeScript-generated Node executables are common?

If not, I'd prefer to leave it as-is, and you can use one of the following solutions to override the icon locally:

## 1. Use your stylesheet
~~~less
.icon[data-name$=".ts"]::before{
	font-family: file-icons;
	content: "\2a6";
	color: #6a9fb5;
}
~~~


## 2. Disable shebang recognition
Go to: **Settings** ⇒ **Packages** ⇒ **`file-icons`** ⇒ **Match strategies** and uncheck the second option:

<img src="https://cloud.githubusercontent.com/assets/2346707/26535715/3bb09006-4473-11e7-9809-608267e99ace.png" width="505" alt="Figure 1" />


## 3. Add a modeline
~~~js
#!/usr/local/bin/node      # -*- TypeScript -*-
~~~
This works because modelines take priority over shebangs when determining which icon to use.

**Comment by sompylasar** [comment]
`Match strategies` looks the right way to go.

**Comment by sompylasar** [comment]
> Do you know if TypeScript-generated Node executables are common?

TypeScript isn't yet common, so no, they aren't **yet**.

**Comment by Alhadis** [comment]
I see. Well, thanks for bringing this up. The package currently lacks heuristic support for icon-assignment (e.g., exceptions to the usual matching methods), but it may be worth adding in future. I'm sure there are other precompiled languages which work the same way as what you've described.

BTW, the hashbang will also affect the file's language classification on GitHub. You can actually fix it both here and in Atom by adding a [`.gitattributes` file to your project](https://github.com/github/linguist#using-gitattributes):

~~~
integration-test.js linguist-language=TypeScript
~~~

I actually recommend this instead, which I forgot about while writing out all that crap above...

**Comment by Alhadis** [comment]
@sompylasar I've made a [unique exception](https://github.com/file-icons/atom/commit/0a36117743da7b8d3272d8544735292778175574) for TypeScript files. This isn't something we normally do, but I've been unable to push this issue out of my head. Specifically, how ridiculous the workarounds are, and how common this scenario might become in future.

I should add that the reason I"m being conservative with icon-matching logic is because the routines responsible for matching an icon to a file potentially run many thousands of times at once... the code needs to be as performant as possible. Hence why I've implemented this as an ad-hoc rule, instead of revising the config structure to accommodate precompilers.
